prereise/gather/griddata/transmission/geometry.py

```diff
@@ -0,0 +1,282 @@
+            "ac": _euclidian(self.a[0], (self.c[0][0], -self.c[0][1])),  # a -> c'
+            "bc": _euclidian(self.b[0], (self.c[0][0], -self.c[0][1])),  # b -> c'
+        }
+        self.equivalent_reflected_distance = _geometric_mean(
+            self.reflected_distance.values()
+        )
+        # Single circuit, so we don't have distances between bundles of the same phase
+        self.phase_self_distances = None
+
+    def calculate_multi_circuit_distances(self):
+        self.phase_self_distances = [
+            _geometric_mean(
+                [
+                    _euclidian(p0, p1)
+                    for i, p0 in enumerate(phase)
+                    for j, p1 in enumerate(phase)
+                    if i > j
+                ]
+            )
+            for phase in (self.a, self.b, self.c)
+        ]
+        # Multi circuit, so we assume tall tower negligible impact from reflectance
+        self.equivalent_reflected_distance = 2 * self.equivalent_height
+
+
+@dataclass
+class Tower:
+    """Given the geometry of a transmission tower and conductor bundle information,
+    estimate per-kilometer inductance, resistance, and shunt capacitance.
+
+    :param PhaseLocations locations: the locations of each conductor bundle.
+    :param ConductorBundle bundle: the parameters of each conductor bundle.
+    """
+
+    locations: PhaseLocations
+    bundle: ConductorBundle
+
+    def __post_init__(self):
+        self.resistance = self.bundle.resistance_per_km / self.locations.circuits
+        self.inductance = self.calculate_inductance_per_km()
+        self.capacitance = self.calculate_shunt_capacitance_per_km()
+
+    def calculate_inductance_per_km(self):
+        denominator = _circuit_bundle_distances(
+            self.bundle.spacing_L, self.locations.phase_self_distances
+        )
+        inductance_per_km = (
+            mu_0 / (2 * pi) * log(self.locations.equivalent_distance / denominator)
+        )
+        return inductance_per_km
+
+    def calculate_shunt_capacitance_per_km(self):
+        denominator = _circuit_bundle_distances(
+            self.bundle.spacing_C, self.locations.phase_self_distances
+        )
+        capacitance_per_km = (2 * pi * epsilon_0) / (
+            log(self.locations.equivalent_distance / denominator)
+            - log(
+                self.locations.equivalent_reflected_distance
+                / (2 * self.locations.equivalent_height)
+            )
+        )
+        return capacitance_per_km
+
+
+def _euclidian(a, b):
+    """Calculate the euclidian distance between two points."""
+    try:
+        if len(a) != len(b):
+            raise ValueError("Length of a and b must be equivalent")
+    except TypeError:
+        raise TypeError(
+            "a and b must both be iterables compatible with the len() function"
+        )
+    return sqrt(sum((x - y) ** 2 for x, y in zip(a, b)))
+
+
+def _geometric_mean(a):
```

Comment:
FYI starting in python 3.8, we could use `from statistics import geometric_mean`